File: Benchmarks/Core/Systems/Contract/ISystem.cs
using Benchmarks.Core.Components;
using Benchmarks.Core.Entity;
using Benchmarks.Core.Scene;
using Benchmarks.Core.Types;

namespace Benchmarks.Core.Systems.Contract
{
    /// <summary>
    /// Define una clase abstracta comun para los sistemas del motor.
    /// Replica exacta del pattern de ISystem del engine.
    /// </summary>
    public abstract class ISystem : IDisposable
    {
        public EntityManager EntityManager { get; set; }

        public virtual void InitializeSystem(EntityManager @entities) { }
        public virtual void SharedEntityInitialize(ref Yotsuba @Entidad) { }
        public virtual void UpdateSystem(GameTime @gameTime) { }
        public virtual void SharedEntityForEachUpdate(ref Yotsuba @Entidad, GameTime time) { }

        public virtual void Dispose() { }

        // --- Entities ---
        public Span<Yotsuba> GetEntitiesAsSpan() => EntityManager.YotsubaEntities.AsSpan();

        public ref Yotsuba GetEntity(int entityId) => ref EntityManager.YotsubaEntities[entityId];

        public ref Yotsuba GetEntityByNameOrDefault(string name) =>
            ref EntityManager.YotsubaEntities.Find(x => x.Name == name);

        // --- Transform ---
        public Span<TransformComponent> GetTransformComponentsAsSpan() => EntityManager.TransformComponents.AsSpan();
        public ref TransformComponent GetTransform2DComponent(int entityId) => ref EntityManager.TransformComponents[entityId];
        public ref TransformComponent GetTransform2DComponent(ref Yotsuba entity) => ref EntityManager.TransformComponents[entity.Id];
        public ref TransformComponent GetTransform2DComponent(Yotsuba entity) => ref EntityManager.TransformComponents[entity.Id];

        // --- Sprite 2D ---
        public Span<SpriteComponent2D> GetSpriteComponentsAsSpan() => EntityManager.Sprite2DComponents.AsSpan();
        public ref SpriteComponent2D GetSprite2DComponent(int entityId) => ref EntityManager.Sprite2DComponents[entityId];
        public ref SpriteComponent2D GetSprite2DComponent(ref Yotsuba entity) => ref EntityManager.Sprite2DComponents[entity.Id];
        public ref SpriteComponent2D GetSprite2DComponent(Yotsuba entity) => ref EntityManager.Sprite2DComponents[entity.Id];

        // --- TileMap 2D ---
        public Span<TileMapComponent2D> GetTilemapComponentsAsSpan() => EntityManager.TileMapComponent2Ds.AsSpan();
        public ref TileMapComponent2D GetTilemapComponent(int entityId) => ref EntityManager.TileMapComponent2Ds[entityId];
        public ref TileMapComponent2D GetTilemapComponent(ref Yotsuba entity) => ref EntityManager.TileMapComponent2Ds[entity.Id];

        // --- Shader ---
        public Span<ShaderComponent> GetShaderComponentsAsSpan() => EntityManager.ShaderComponents.AsSpan();
        public ref ShaderComponent GetShaderComponent(int entityId) => ref EntityManager.ShaderComponents[entityId];
        public ref ShaderComponent GetShaderComponent(ref Yotsuba entity) => ref EntityManager.ShaderComponents[entity.Id];

        // --- Script ---
        public Span<ScriptComponent> GetScriptComponentsAsSpan() => EntityManager.ScriptComponents.AsSpan();
        public ref ScriptComponent GetScriptComponent(int entityId) => ref EntityManager.ScriptComponents[entityId];

        // --- RigidBody 2D ---
        public Span<RigidBodyComponent2D> GetRigidBodyComponentsAsSpan() => EntityManager.Rigidbody2DComponents.AsSpan();
        public ref RigidBodyComponent2D GetRigidBodyComponent(int entityId) => ref EntityManager.Rigidbody2DComponents[entityId];
        public ref RigidBodyComponent2D GetRigidBodyComponent(ref Yotsuba entity) => ref EntityManager.Rigidbody2DComponents[entity.Id];

        // --- Button 2D ---
        public Span<ButtonComponent2D> GetButtonComponentsAsSpan() => EntityManager.Button2DComponents.AsSpan();
        public ref ButtonComponent2D GetButtonComponent(int entityId) => ref EntityManager.Button2DComponents[entityId];

        // --- Model 3D ---
        public Span<ModelComponent3D> GetModelsComponentsAsSpan() => EntityManager.ModelComponents3D.AsSpan();
        public ref ModelComponent3D GetModelComponent(int entityId) => ref EntityManager.ModelComponents3D[entityId];

        // --- YTB Model 3D ---
        public Span<YTBModelComponent3D> GetYTBModelComponentsAsSpan() => EntityManager.YtbModelComponents.AsSpan();
        public ref YTBModelComponent3D GetYTBModelComponent(int entityId) => ref EntityManager.YtbModelComponents[entityId];

        // --- Input ---
        public Span<InputComponent> GetInputComponentsAsSpan() => EntityManager.InputComponents.AsSpan();
        public ref InputComponent GetInputComponent(int entityId) => ref EntityManager.InputComponents[entityId];
        public ref InputComponent GetInputComponent(ref Yotsuba entity) => ref EntityManager.InputComponents[entity.Id];

        // --- Font 2D ---
        public Span<FontComponent2D> GetFontComponentsAsSpan() => EntityManager.Font2DComponents.AsSpan();
        public ref FontComponent2D GetFontComponent(int entityId) => ref EntityManager.Font2DComponents[entityId];

        // --- Animation 2D ---
        public Span<AnimationComponent2D> GetAnimationComponentsAsSpan() => EntityManager.Animation2DComponents.AsSpan();
        public ref AnimationComponent2D GetAnimationComponent(int entityId) => ref EntityManager.Animation2DComponents[entityId];
        public ref AnimationComponent2D GetAnimationComponent(ref Yotsuba entity) => ref EntityManager.Animation2DComponents[entity.Id];

        public static bool IsDefaultEntity(ref Yotsuba entity) => entity.Equals(default(Yotsuba));
    }
}

File: Benchmarks_FSharp/Core/Systems/Contract/ISystem.fs
namespace Benchmarks.Core.Systems.Contract

open System
open Benchmarks.Core.Components
open Benchmarks.Core.Entity
open Benchmarks.Core.Scene
open Benchmarks.Core.Types

[<AbstractClass>]
type ISystem() =
    member val EntityManager: EntityManager = Unchecked.defaultof<EntityManager> with get, set

    abstract InitializeSystem: EntityManager -> unit
    default _.InitializeSystem(_) = ()

    abstract SharedEntityInitialize: entity: byref<Yotsuba> -> unit
    default _.SharedEntityInitialize(_) = ()

    abstract UpdateSystem: GameTime -> unit
    default _.UpdateSystem(_) = ()

    abstract SharedEntityForEachUpdate: entity: byref<Yotsuba> * time: GameTime -> unit
    default _.SharedEntityForEachUpdate(_, _) = ()

    abstract Dispose: unit -> unit
    default _.Dispose() = ()

    interface IDisposable with
        member this.Dispose() = this.Dispose()

    // --- Entities ---
    member this.GetEntitiesAsSpan() = this.EntityManager.YotsubaEntities.AsSpan()
    member this.GetEntity(entityId: int) = &this.EntityManager.YotsubaEntities.[entityId]
    member this.GetEntityByNameOrDefault(name: string) =
        &this.EntityManager.YotsubaEntities.Find(fun x -> x.Name = name)

    // --- Transform ---
    member this.GetTransformComponentsAsSpan() = this.EntityManager.TransformComponents.AsSpan()
    member this.GetTransform2DComponent(entityId: int) = &this.EntityManager.TransformComponents.[entityId]

    // --- Sprite 2D ---
    member this.GetSpriteComponentsAsSpan() = this.EntityManager.Sprite2DComponents.AsSpan()
    member this.GetSprite2DComponent(entityId: int) = &this.EntityManager.Sprite2DComponents.[entityId]

    // --- TileMap 2D ---
    member this.GetTilemapComponentsAsSpan() = this.EntityManager.TileMapComponent2Ds.AsSpan()
    member this.GetTilemapComponent(entityId: int) = &this.EntityManager.TileMapComponent2Ds.[entityId]

    // --- Shader ---
    member this.GetShaderComponentsAsSpan() = this.EntityManager.ShaderComponents.AsSpan()
    member this.GetShaderComponent(entityId: int) = &this.EntityManager.ShaderComponents.[entityId]

    // --- Script ---
    member this.GetScriptComponentsAsSpan() = this.EntityManager.ScriptComponents.AsSpan()
    member this.GetScriptComponent(entityId: int) = &this.EntityManager.ScriptComponents.[entityId]

    // --- RigidBody 2D ---
    member this.GetRigidBodyComponentsAsSpan() = this.EntityManager.Rigidbody2DComponents.AsSpan()
    member this.GetRigidBodyComponent(entityId: int) = &this.EntityManager.Rigidbody2DComponents.[entityId]

    // --- Button 2D ---
    member this.GetButtonComponentsAsSpan() = this.EntityManager.Button2DComponents.AsSpan()
    member this.GetButtonComponent(entityId: int) = &this.EntityManager.Button2DComponents.[entityId]

    // --- Model 3D ---
    member this.GetModelsComponentsAsSpan() = this.EntityManager.ModelComponents3D.AsSpan()
    member this.GetModelComponent(entityId: int) = &this.EntityManager.ModelComponents3D.[entityId]

    // --- YTB Model 3D ---
    member this.GetYTBModelComponentsAsSpan() = this.EntityManager.YtbModelComponents.AsSpan()
    member this.GetYTBModelComponent(entityId: int) = &this.EntityManager.YtbModelComponents.[entityId]

    // --- Input ---
    member this.GetInputComponentsAsSpan() = this.EntityManager.InputComponents.AsSpan()
    member this.GetInputComponent(entityId: int) = &this.EntityManager.InputComponents.[entityId]

    // --- Font 2D ---
    member this.GetFontComponentsAsSpan() = this.EntityManager.Font2DComponents.AsSpan()
    member this.GetFontComponent(entityId: int) = &this.EntityManager.Font2DComponents.[entityId]

    // --- Animation 2D ---
    member this.GetAnimationComponentsAsSpan() = this.EntityManager.Animation2DComponents.AsSpan()
    member this.GetAnimationComponent(entityId: int) = &this.EntityManager.Animation2DComponents.[entityId]

    static member IsDefaultEntity(entity: byref<Yotsuba>) = entity.Equals(Unchecked.defaultof<Yotsuba>)
